machineconfiguration/v1/types_machineosbuild.go

```diff
@@ -0,0 +1,178 @@
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineOSBuildList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineOSBuild `json:"items"`
+}
+
+// MachineOSBuildSpec describes information about a build process primarily populated from a MachineOSConfig object.
+type MachineOSBuildSpec struct {
+	// desiredConfig is the desired config we want to build an image for.
+	// +kubebuilder:validation:Required
+	DesiredConfig RenderedMachineConfigReference `json:"desiredConfig"`
+	// machineOSConfig is the config object which the build is based off of
+	// +kubebuilder:validation:Required
+	MachineOSConfig MachineOSConfigReference `json:"machineOSConfig"`
+	// renderedImagePushspec is set from the MachineOSConfig
+	// The format of the image pullspec is:
+	// host[:port][/namespace]/name:<tag> or svc_name.namespace.svc[:port]/repository/name:<tag>
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=447
+	// +kubebuilder:validation:XValidation:rule=`self.matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?(/[a-zA-Z0-9-_]{1,61})*/[a-zA-Z0-9-_.]+:[a-zA-Z0-9._-]+$') || self.matches('^[^.]+\\.[^.]+\\.svc:\\d+\\/[^\\/]+\\/[^\\/]+:[^\\/]+$')`,message="the OCI Image name should follow the host[:port][/namespace]/name format, resembling a valid URL without the scheme. Or it must be a valid .svc followed by a port, repository, image name, and tag."
+	// +kubebuilder:validation:Required
+	RenderedImagePushspec string `json:"renderedImagePushspec"`
+}
+
+// MachineOSBuildStatus describes the state of a build and other helpful information.
+type MachineOSBuildStatus struct {
+	// conditions are state related conditions for the build. Valid types are:
+	// Prepared, Building, Failed, Interrupted, and Succeeded
+	// once a Build is marked as Failed, no future conditions can be set. This is enforced by the MCO.
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// BuilderReference describes which ImageBuilder backend to use for this build
+	// +optional
+	BuilderReference *MachineOSBuilderReference `json:"builderReference"`
+	// relatedObjects is a list of objects that are related to the build process.
+	// +kubebuilder:validation:MaxItems=10
+	// +listType=map
+	// +listMapKey=name
+	// +listMapKey=resource
+	RelatedObjects []ObjectReference `json:"relatedObjects,omitempty"`
+	// buildStart describes when the build started.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="buildStart is immutable once set"
+	// +kubebuilder:validation:Required
+	BuildStart metav1.Time `json:"buildStart"`
+	// buildEnd describes when the build ended.
```

Comment:
The build has not yet completed. The end time is added when the build completes, fails, or is interrupted.

Reply:
```suggestion
	// buildEnd describes when the build ended.
	// When omitted the build is still in progress.
	// It will be populated once the build completes, fails or is interrupted.
```

Reply:
Is there a dependency between start time and end time? We could validate that buildEnd is only added after buildStart, and we could even validate that buildEnd is after buildStart 🤔